core/cmake/vecmem-setup-core.cmake

```diff
@@ -0,0 +1,109 @@
+# VecMem project, part of the ACTS project (R&D line)
+#
+# (c) 2024 CERN for the benefit of the ACTS project
+#
+# Mozilla Public License Version 2.0
+
+# VecMem include(s).
+include( vecmem-check-language )
+include( vecmem-check-sycl-source-compiles )
+
+# CMake include(s).
+include( CheckCXXSourceCompiles )
+
+# Helper function setting up the public definitions for vecmem::core
+#
+# Usage: vecmem_setup_core( vecmem::core )
+#
+function( vecmem_setup_core libName )
+
+   # Check if vecmem::posix_device_atomic_ref is usable.
+   get_target_property( _coreIncludes "${libName}"
+      INTERFACE_INCLUDE_DIRECTORIES )
+   set( CMAKE_REQUIRED_INCLUDES ${_coreIncludes} )
+   unset( _coreIncludes )
+   check_cxx_source_compiles( "
+      #include <vecmem/memory/details/posix_device_atomic_ref.hpp>
+      int main() {
+         int foo = 0;
+         vecmem::posix_device_atomic_ref<int> ref{foo};
+         return 0;
+      }
+      " VECMEM_SUPPORT_POSIX_ATOMIC_REF )
+   if( VECMEM_SUPPORT_POSIX_ATOMIC_REF )
+      target_compile_definitions( ${libName} INTERFACE
+         $<BUILD_INTERFACE:VECMEM_SUPPORT_POSIX_ATOMIC_REF> )
+   endif()
+   unset( CMAKE_REQUIRED_INCLUDES )
+
+   # Check if a SYCL compiler is available.
+   vecmem_check_language( SYCL )
+
+   # Set the SYCL definitions only if the build system knows about SYCL.
+   if( CMAKE_SYCL_COMPILER )
+
+      # Test which SYCL printf function(s) is/are available.
```

Comment:
Ouff, that's quite a mess. But I guess it has to be done.

Reply:
This is the logic we had in place for a number of years by now. Unfortunately `printf(...)` is a oneAPI extension to the SYCL standard. The logging functionality would need to be seriously redesigned to make it work with [sycl::stream](https://www.intel.com/content/www/us/en/docs/oneapi/optimization-guide-gpu/2023-0/doing-io-in-the-kernel.html). Since I was able to use for instance [AdaptiveCpp](https://github.com/AdaptiveCpp/AdaptiveCpp), which does not support `printf(...)`, by just not turning on any debug messages, it was fine so far to only get debugging output with oneAPI.